pkg/controller/elasticsearch/pdb/reconcile_with_roles.go

```diff
@@ -0,0 +1,449 @@
+		nodeRoles[i] = esv1.NodeRole(role)
+	}
+	return nodeRoles, nil
+}
+
+// createPDBForStatefulSets creates a PDB for a group of StatefulSets with shared roles.
+func createPDBForStatefulSets(
+	es esv1.Elasticsearch,
+	// role is the role used to determine the maxUnavailable value.
+	role esv1.NodeRole,
+	// roleName is used to determine the name of the PDB.
+	roleName string,
+	// statefulSets are the statefulSets grouped into this pdb.
+	statefulSets []appsv1.StatefulSet,
+	// allStatefulSets are all statefulSets in the whole ES cluster.
+	allStatefulSets sset.StatefulSetList,
+	meta metadata.Metadata,
+) (*policyv1.PodDisruptionBudget, error) {
+	if len(statefulSets) == 0 {
+		return nil, nil
+	}
+
+	pdb := &policyv1.PodDisruptionBudget{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      esv1.PodDisruptionBudgetNameForRole(es.Name, roleName),
+			Namespace: es.Namespace,
+		},
+		Spec: buildRoleSpecificPDBSpec(es, role, statefulSets, allStatefulSets),
+	}
+
+	mergedMeta := meta.Merge(metadata.Metadata{
+		Labels:      pdb.Labels,
+		Annotations: pdb.Annotations,
+	})
+	pdb.Labels = mergedMeta.Labels
+	pdb.Annotations = mergedMeta.Annotations
+
+	// Set owner reference
+	if err := controllerutil.SetControllerReference(&es, pdb, scheme.Scheme); err != nil {
+		return nil, err
+	}
+
+	return pdb, nil
+}
+
+// buildRoleSpecificPDBSpec returns a policyv1.PodDisruptionBudgetSpec for a specific node role.
+func buildRoleSpecificPDBSpec(
+	es esv1.Elasticsearch,
+	role esv1.NodeRole,
+	// statefulSets are the statefulSets grouped into this pdb.
+	statefulSets sset.StatefulSetList,
+	// allStatefulSets are all statefulSets in the whole ES cluster.
+	allStatefulSets sset.StatefulSetList,
+) policyv1.PodDisruptionBudgetSpec {
+	// Get the allowed disruptions for this role based on cluster health and role type
+	allowedDisruptions := allowedDisruptionsForRole(es, role, allStatefulSets)
+
+	spec := policyv1.PodDisruptionBudgetSpec{
+		MaxUnavailable: &intstr.IntOrString{Type: intstr.Int, IntVal: allowedDisruptions},
+	}
+
+	// Get StatefulSet names for the selector
+	ssetNames := make([]string, 0, len(statefulSets))
+	for _, sset := range statefulSets {
+		ssetNames = append(ssetNames, sset.Name)
+	}
+
+	// Sort for consistency
+	sort.Strings(ssetNames)
+
+	spec.Selector = selectorForStatefulSets(es, ssetNames)
+	return spec
+}
+
+// allowedDisruptionsForRole returns the maximum number of pods that can be disrupted for a given role.
+// Notes:
+//   - Disruptions are never allowed when the cluster health is unknown, empty or red, even if the
+//     cluster is not a highly available cluster. This is to ensure no further damage can be done
+//     to a cluster which is already unhealthy for an unknown reason.
+//   - In a single node cluster (not highly-available) always allow 1 disruption
+//     to ensure K8s nodes operations can be performed when the cluster health is not unknown or red.
+//   - For data roles, allow 1 disruption if health is green.
```

Comment:
One thing which is not obvious when reading this code is "when" frozen is handled differently. Let's maybe try to explain where the magic happens in the godoc?
```suggestion
//   - For data roles, excluding frozen which should have been filtered in groupBySharedRoles, allow 1 disruption if health is green.
```